sno/working_copy/sqlserver.py

```diff
@@ -0,0 +1,553 @@
+            ON SRC."table_name" = STA."table_name" AND SRC."key" = STA."key"
+            WHEN MATCHED THEN
+                UPDATE SET "value" = SRC."value"
+            WHEN NOT MATCHED THEN
+                INSERT ("table_name", "key", "value") VALUES (SRC."table_name", SRC."key", SRC."value");
+            """,
+            {"value": tree_id},
+        )
+        return r.rowcount
+
+    def _write_meta(self, sess, dataset):
+        """Write the title (as a comment) and the CRS. Other metadata is not stored in a PostGIS WC."""
+        self._write_meta_title(sess, dataset)
+
+    def _write_meta_title(self, sess, dataset):
+        """Write the dataset title as a comment on the table."""
+        # TODO - probably need to use sp_addextendedproperty @name=N'MS_Description'
+        pass
+
+    def delete_meta(self, dataset):
+        """Delete any metadata that is only needed by this dataset."""
+        pass  # There is no metadata except for the spatial_ref_sys table.
+
+    def _create_spatial_index(self, sess, dataset):
+        L = logging.getLogger(f"{self.__class__.__qualname__}._create_spatial_index")
+
+        geom_col = dataset.geom_column_name
+
+        # Create the SQL Server Spatial Index
+        L.debug("Creating spatial index for %s.%s", dataset.table_name, geom_col)
+        t0 = time.monotonic()
+        index_name = f"{dataset.table_name}_idx_{geom_col}"
+        sess.execute(
+            f"""
+            CREATE SPATIAL INDEX {self.quote(index_name)}
+            ON {self.table_identifier(dataset)} ({self.quote(geom_col)});
+            """
+        )
+        L.info("Created spatial index in %ss", time.monotonic() - t0)
+
+    def _drop_spatial_index(self, sess, dataset):
+        # SQL server deletes the spatial index automatically when the table is deleted.
+        pass
+
+    def _quoted_trigger_name(self, dataset):
+        trigger_name = f"{dataset.table_name}_sno_track"
+        return f"{self.DB_SCHEMA}.{self.quote(trigger_name)}"
+
+    def _create_triggers(self, sess, dataset):
+        pk_name = dataset.primary_key
+        escaped_table_name = dataset.table_name.replace("'", "''")
+
+        sess.execute(
+            f"""
+            CREATE TRIGGER {self._quoted_trigger_name(dataset)} ON {self.table_identifier(dataset)}
+            AFTER INSERT, UPDATE, DELETE AS
+            BEGIN
+                MERGE {self.SNO_TRACK} TRA
+                USING
+                    (SELECT '{escaped_table_name}', {self.quote(pk_name)} FROM inserted
+                    UNION SELECT '{escaped_table_name}', {self.quote(pk_name)} FROM deleted)
+                    AS SRC (table_name, pk)
+                ON SRC.table_name = TRA.table_name AND SRC.pk = TRA.pk
+                WHEN NOT MATCHED THEN INSERT (table_name, pk) VALUES (SRC.table_name, SRC.pk);
+            END;
+            """,
+            {"table_name": dataset.table_name},
+        )
```

Comment:
Can we use the SA statement builder to safely assemble the content of the trigger, then serialise it as a string into the outer SA statement?

Reply:
The fstring interpolation was because the database driver won't necessarily accept placeholders everywhere that a string literal would be valid - not sure if this is because of nested statements, or because of triggers in this case. Either way, the eventual SQL that is sent to the database driver must have the string literal inlined, hence the use of f-strings.
(`table_name` being there and unused was an artifact of me trying it one way and then realising I had to do it another, of course).

I've managed to get the sqlalchemy to inline and escape of the string literal itself using some magic.
I think this is about as good as its going to get - sqlalchemy doesn't have built in functions for CREATE TRIGGER or MERGE USING, so its always going to be custom SQL. I can write compiler extensions that generate this custom SQL, but then it probably becomes more likely that I will mess up escaping etc, not less.